local-cluster/tests/local_cluster.rs

```diff
@@ -1876,6 +1876,359 @@ fn test_incremental_snapshot_download() {
+    info!("Waiting for the validator to see enough slots to cross a full snapshot interval...");
+    let starting_slot = incremental_snapshot_archive_info.slot();
+    let timer = Instant::now();
+    loop {
+        let validator_current_slot = cluster
+            .get_validator_client(&validator_identity.pubkey())
+            .unwrap()
+            .get_slot_with_commitment(CommitmentConfig::finalized())
+            .unwrap();
+        if validator_current_slot > (starting_slot + full_snapshot_interval) {
+            break;
+        }
+        assert!(
+            timer.elapsed() < Duration::from_secs(20),
+            "It should not take longer than 20 seconds to cross the next full snapshot interval."
+        );
+        std::thread::yield_now();
+    }
+    trace!("Waited {:?}", timer.elapsed());
+
+    trace!(
+        "Delete all the snapshots on the validator and restore the originals from the backup..."
+    );
+    delete_files(validator_snapshot_test_config.snapshot_archives_dir.path());
+    copy_files(
+        validator_snapshot_backup_dir.path(),
+        validator_snapshot_test_config.snapshot_archives_dir.path(),
+    );
+
+    // Get the highest full snapshot slot *before* restarting, as a comparison
+    let old_validator_highest_full_snapshot_slot =
+        snapshot_utils::get_highest_full_snapshot_archive_slot(
+            validator_snapshot_test_config.snapshot_archives_dir.path(),
+        )
+        .unwrap();
+
+    info!("Restarting the validator...");
+    let validator_info = cluster.exit_node(&validator_identity.pubkey());
+    cluster.restart_node(
+        &validator_identity.pubkey(),
+        validator_info,
+        SocketAddrSpace::Unspecified,
+    );
+
+    // Now, we want to ensure that the validator can make a new incremental snapshot based on the
+    // new full snapshot that was created during the restart.
+    let timer = Instant::now();
+    loop {
+        if let Some(highest_full_snapshot_slot) =
+            snapshot_utils::get_highest_full_snapshot_archive_slot(
+                validator_snapshot_test_config.snapshot_archives_dir.path(),
+            )
+        {
+            if highest_full_snapshot_slot > old_validator_highest_full_snapshot_slot {
+                if let Some(highest_incremental_snapshot_slot) =
+                    snapshot_utils::get_highest_incremental_snapshot_archive_slot(
+                        validator_snapshot_test_config.snapshot_archives_dir.path(),
+                        highest_full_snapshot_slot,
+                    )
+                {
+                    info!("Success! Made new full and incremental snapshots!");
+                    trace!(
+                        "Full snapshot slot: {}, incremental snapshot slot: {}",
+                        highest_full_snapshot_slot,
+                        highest_incremental_snapshot_slot
+                    );
+                    break;
+                }
+            }
+        }
```

Comment:
Wondering if there's a way to verify that the hash of the snapshot that was created at startup here matches the snapshot on the leader slot.  That way we can verify the snapshot was correctly taken.

If we didn't purge the old bank snapshots on the leader, then maybe this would be possible since we could search its snapshot directory for such a snapshot

Unfortunately we do purge here: https://github.com/solana-labs/solana/blob/d0633c5547e4fd7a3099c45aedc6003397ff7db0/runtime/src/snapshot_utils.rs#L1567 😛, hmmm....

Reply:
Yeah, this is a good test to add. Done!